### PR TITLE
Catch nil values for FoundSubjectsByResourceID map and return as errors

### PR DESCRIPTION
**Pull request by josephschorr**
Also ensure we never add nils in the first place

### CODE REVIEW

File: internal/graph/lookupsubjects.go [hunk 1]
```diff
@@ -460,7 +464,10 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
 
 		for _, result := range collector.Results() {
 			metadata = combineResponseMetadata(metadata, result.Metadata)
-			foundSubjects.UnionWith(result.FoundSubjectsByResourceId)
+			err := foundSubjects.UnionWith(result.FoundSubjectsByResourceId)
```
Comment (jakedt) [inline internal/graph/lookupsubjects.go:467]: You should use the `if err := thing(); err != nil` pattern here and at other callsites.

File: internal/graph/lookupsubjects.go [hunk 2]
```diff
@@ -460,7 +464,10 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
...
+			if err != nil {
+				return err
```
Comment (jakedt) [inline internal/graph/lookupsubjects.go:469]: Wrap these errors with higher level context about _why_ you were trying to Union, so we can trace which callsite caused the problem.

File: internal/datasets/subjectsetbyresourceid.go [hunk 3]
```diff
@@ -36,12 +42,22 @@ func (ssr SubjectSetByResourceID) AddFromRelationship(relationship *core.Relatio
 }
 
 // UnionWith unions the map's sets with the other map of sets provided.
-func (ssr SubjectSetByResourceID) UnionWith(other map[string]*v1.FoundSubjects) {
+func (ssr SubjectSetByResourceID) UnionWith(other map[string]*v1.FoundSubjects) error {
```
Comment (jakedt) [inline internal/datasets/subjectsetbyresourceid.go:45]: Can you create a test which reproduces why this happens in real running-systems to verify the fix?

**Review by jakedt — CHANGES_REQUESTED** [review]

**Comment by josephschorr** [comment]
Updated

File: internal/datasets/subjectsetbyresourceid.go [hunk 4]
```diff
@@ -18,30 +20,44 @@ type SubjectSetByResourceID struct {
...
 	for resourceID, subjects := range other {
+		if subjects == nil {
+			return fmt.Errorf("received nil FoundSubjects in other map of SubjectSetByResourceID's UnionWith for key %s", resourceID)
+		}
+
 		for _, subject := range subjects.FoundSubjects {
-			ssr.add(resourceID, subject)
+			err := ssr.add(resourceID, subject)
```
Comment (jakedt) [inline internal/datasets/subjectsetbyresourceid.go:53]: `if err := ...; err != nil`

**Review by jakedt — APPROVED** [review]

**Pull request merged**